### DISCUSSION TITLE
Channel specific calibration before categorization

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#546] [General]
Hello CF Experts,

I am working on H to tautau CP violation analysis for four different final states. All of the four channels require different working points for the tau energy scale correction. The `channel_ids` are being produced later at the end of selection. Therefore, it is probably not possible to use the channel specific tauID WPs before knowing the categories.

If there is any workaround, I would request you to let me know.

There is some bypass trick that I am following currently:

 - [Here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/calibration/tau.py#L26-L30), I am doing the same thing for different WPs
 - [Here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/selection/main.py#L244-L250), I am doing the tau-selection three times 
 - Then I used respective taus to build the lepton pairs e.g. e-tau, mu-tau or tau-tau, like [this](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/selection/main.py#L260-L283), and I am using the calibrated properties as mentioned [here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/selection/lepton_pair_tautau.py#L281-L285)
 - After creating at least one lep-pair per event, I also want to save the eles, muons and taus in the pairs as the `events.Muon`, `events.Electron` and `events.Tau`. I am saving other collections as well, like [here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/selection/physics_objects.py#L490-L494)
 - [Here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/selection/higgscand.py#L247), I am emplacing the calibrated taus back to the original tau array by their `rawIdx`, as described [here](https://github.com/gsaha009/CPinHToTauTau/blob/dev/httcp/calibration/tau.py#L147-L265) -- [related awkward discussion : [link](https://github.com/scikit-hep/awkward/discussions/3250)]
 
 By this bypass, I have managed to run all channels in one go. But, I was thinking if I want to add up and down variation for tec, it might get more complicated.
 
 Any type of suggestion would be very helpful. 

Thank you :)

**Comment by pkausw** [comment]
Hi @gsaha009 ,

quick question - what keeps you from using the `channel_ids` producer in your Calibrator? I guess this would require some minimal/common calibration of your objects first, in principle you should be able to do something like this I think

```python
# assuming you have other calibrators for your objects
from YOUR_ANALYSIS import tes_calibration_minimal, tes_calibration_etau, tes_calibration_mutau, tes_calibration_tautau
from columnflow.production import channel_ids
from columnflow.calibration import Calibrator, calibrator

@calibrator(
    uses={channel_ids, tes_calibration_minimal, tes_calibration_etau, tes_calibration_mutau, tes_calibration_tautau},
    produces={channel_ids, tes_calibration_minimal, tes_calibration_etau, tes_calibration_mutau, tes_calibration_tautau},
)
def tes_calibrator(self, events, **kwargs):
    # first do the minimal calibration in case your categories/channels depend on kinematic regions
    events = self[tes_calibration_minimal](events, **kwargs)
    # now create the channel_ids
    events = self[channel_ids](events, **kwargs)

    # now do the calibration for the individual channels. Probably advisable to use a ak.where when setting the columns in events
    events = self[test_calibration_etau](events, **kwargs)
    events = self[test_calibration_mutau](events, **kwargs)
    events = self[test_calibration_tautau](events, **kwargs)

    return events
```

This way, you insert the correct calibration for the individual events. This means that you should also be able to use the correct shift for these events when doing the replacement - provided that the name of these shifts is the same for the individual channels (i.e. you correlate the uncertainties in the fit). But I think this should work even if the names are different - you'll probably just need to properly define the column aliases in your config objects.

Cheers,
Philip
